packages/vite/src/node/build.ts

```diff
@@ -422,18 +422,19 @@ async function doBuild(
   try {
     const buildOutputOptions = (output: OutputOptions = {}): OutputOptions => {
       const cjsSsrBuild = ssr && config.ssr?.format === 'cjs'
+      const format = output.format || cjsSsrBuild ? 'cjs' : 'es'
       return {
         dir: outDir,
         // Default format is 'es' for regular and for SSR builds
-        format: cjsSsrBuild ? 'cjs' : 'es',
+        format,
         exports: cjsSsrBuild ? 'named' : 'auto',
         sourcemap: options.sourcemap,
         name: libOptions ? libOptions.name : undefined,
         generatedCode: 'es2015',
         entryFileNames: ssr
-          ? `[name].js`
+          ? `[name].${resolveOutputJsExtensionFromRoot(format, config.root)}`
           : libOptions
-          ? resolveLibFilename(libOptions, output.format || 'es', config.root)
+          ? resolveLibFilename(libOptions, format, config.root)
           : path.posix.join(options.assetsDir, `[name].[hash].js`),
         chunkFileNames: libOptions
           ? `[name].[hash].js`
```

Comment:
@sachinraja @antfu shouldn't chunks also have a derived `mjs`/`cjs`/`js` extension for chunks in lib mode (and also SSR?)

Reply:
To me I think it's better to consist the extension for both entry and chunks

Reply:
Yes, they must to be compliant with node resolution

Reply:
Done, it should now work for SSR and chunks